File: videos/story.py
from pathlib import Path
from dataclasses import dataclass
import re
import json

# Token groups for German language start with G, for Russian with R
TOKEN_GROUPS_REGEX = re.compile(r"\([GR]\d+(,[GR]\d+)*\)")

VALID_LEVELS = [
    "A1",
    "A1-A2",
    "A2",
    "A2-B1",
    "B1",
    "B1-B2",
    "B2",
    "B2-C1",
    "C1",
]


@dataclass
class Token:
    text: str
    group_ids: list[str]

    def to_str(self) -> str:
        if len(self.group_ids) == 0:
            return self.text
        groups = ",".join(self.group_ids)
        return f"{self.text}({groups})"


@dataclass
class Segment:
    tokens: list[Token]
    index: int

    def to_str(self) -> str:
        return " ".join(token.to_str() for token in self.tokens)

    def to_str_no_groups(self) -> str:
        return " ".join(token.text for token in self.tokens)


@dataclass
class Sentence:
    segments: list[Segment]
    index: int

    def to_plain_str(self) -> str:
        return " ".join(segment.to_str_no_groups() for segment in self.segments)


@dataclass
class Scene:
    sentences: list[Sentence]
    image_id: str | None


@dataclass
class Paragraph:
    scenes: list[Scene]
    image_id: str | None


@dataclass
class Chapter:
    title: str | None
    paragraphs: list[Paragraph]


@dataclass
class Story:
    title: str
    chapters: list[Chapter]
    image_id: str | None


@dataclass
class StoryMultilingual:
    id: str
    level: str
    localizations: dict[str, Story]


def _parse_sentence(line: str) -> Sentence:
    segments: list[Segment] = []
    line = line.replace("@", "\n")
    for segment in line.split("||"):
        segment = segment.rstrip()
        tokens: list[Token] = []
        for token in segment.split(" "):
            group_ids: list[str] = []
            match = TOKEN_GROUPS_REGEX.search(token)
            if match:
                group_ids = match.group().strip("()").split(",")
            text = re.sub(TOKEN_GROUPS_REGEX, "", token)
            tokens.append(Token(text=text, group_ids=group_ids))
        segments.append(Segment(tokens, index=0))
    return Sentence(segments, index=0)


def _recalculate_sentence_and_segment_indexes(story: StoryMultilingual) -> None:
    for story_localization in story.localizations.values():
        sentence_idx = 0
        segment_idx = 0
        for chapter in story_localization.chapters:
            for paragraph in chapter.paragraphs:
                for scene in paragraph.scenes:
                    for sentence in scene.sentences:
                        sentence.index = sentence_idx
                        sentence_idx += 1
                        for segment in sentence.segments:
                            segment.index = segment_idx
                            segment_idx += 1


def parse_story(lines: list[str]) -> StoryMultilingual:
    locales: list[str] = []
    current_locale_idx = 0
    stories: list[Story] = []
    story_id: str | None = None
    level: str | None = None

    def next_story_locale_idx() -> int:
        nonlocal current_locale_idx
        current_locale_idx = (current_locale_idx + 1) % len(locales)
        return current_locale_idx

    def start_stories(titles: list[str], image_id: str | None):
        assert len(titles) == len(locales)
        for title in titles:
            stories.append(Story(title=title, chapters=[], image_id=image_id))

    def ensure_chapters():
        for story in stories:
            if not story.chapters:
                story.chapters.append(Chapter(title=None, paragraphs=[]))

    def start_chapter(titles: list[str | None]):
        assert len(titles) == len(stories)
        for story, title in zip(stories, titles):
            story.chapters.append(Chapter(title=title, paragraphs=[]))

    def ensure_paragraphs():
        ensure_chapters()
        for story in stories:
            chapter = story.chapters[-1]
            if not chapter.paragraphs:
                chapter.paragraphs.append(Paragraph(scenes=[], image_id=None))

    def start_paragraph(image_id: str | None):
        ensure_chapters()
        for story in stories:
            chapter = story.chapters[-1]
            chapter.paragraphs.append(Paragraph(scenes=[], image_id=image_id))

    def ensure_scenes():
        ensure_paragraphs()
        for story in stories:
            chapter = story.chapters[-1]
            paragraph = chapter.paragraphs[-1]
            if not paragraph.scenes:
                paragraph.scenes.append(Scene(sentences=[], image_id=None))

    def start_scene(image_id: str | None):
        ensure_paragraphs()
        for story in stories:
            chapter = story.chapters[-1]
            paragraph = chapter.paragraphs[-1]
            paragraph.scenes.append(Scene(sentences=[], image_id=image_id))

    last_tag = ""
    for line in lines:
        try:
            line = line.strip()

            if line.startswith("#") or not line:
                continue

            if not story_id:
                story_id, level = line.split(maxsplit=1)
                assert level in VALID_LEVELS
                continue

            if not locales:
                locales = [l.strip() for l in line.split(",")]
                continue

            tag = line.split()[0] if line.startswith("/") else ""
            if tag != last_tag:
                assert current_locale_idx == 0, f"Line: {line}"
                last_tag = tag
            value = line[len(tag) :].strip()
            params = {}
            if tag and value:
                params = json.loads(value)

            if tag == "/t":
                start_stories(params["titles"], params.get("image_id"))
                continue

            if tag == "":
                ensure_scenes()
                sentence = _parse_sentence(value)
                stories[current_locale_idx].chapters[-1].paragraphs[-1].scenes[
                    -1
                ].sentences.append(sentence)
                next_story_locale_idx()
                continue

            if tag == "/s":
                start_scene(params.get("image_id"))
                continue

            if tag == "/p":
                start_paragraph(params.get("image_id"))
                continue

            if tag == "/c":
                titles = params.get("titles")
                if not titles:
                    titles = [None] * len(locales)
                start_chapter(titles)
                continue
        except Exception as e:
            print(f"Error on line: {line}")
            raise e

    assert current_locale_idx == 0
    assert story_id is not None
    assert level is not None

    story = StoryMultilingual(
        id=story_id,
        level=level,
        localizations={
            locale: story for locale, story in zip(locales, stories)
        },
    )
    _recalculate_sentence_and_segment_indexes(story)
    return story


if __name__ == "__main__":
    test_story_path = Path(
        "/mnt/hgfs/shared/data/stories/015-beneath-paint/C1/mapping_grouped_new_format.txt"
    )
    with open(test_story_path, "r", encoding="utf-8") as file:
        all_lines = file.read().splitlines()
    parsed_story = parse_story(all_lines)
    print(
        json.dumps(
            parsed_story,
            indent=1,
            ensure_ascii=False,
            default=lambda o: o.__dict__,
        )
    )

File: api/src/story/story.go
package story

import (
	"encoding/json"
	"errors"
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// Example: "(1)", "(1, 22)"
var TOKEN_GROUPS_REGEX = regexp.MustCompile(`\(\d+(,\d+)*\)`)

var LEVELS = []string{
	"A1", "A1-A2",
	"A2", "A2-B1",
	"B1", "B1-B2",
	"B2", "B2-C1",
	"C1",
}

type Id = string
type Level = string
type Locale = string

type StoryDescriptor struct {
	Id      Id       `json:"id"`
	Level   Level    `json:"level"`
	Locales []Locale `json:"locales"`
	Titles  []string `json:"titles"` // Titles in the order of locales
}

type Token struct {
	Text     string `json:"text"`
	GroupIds []int  `json:"group_ids,omitempty"`
}

func (t Token) ToStr() string {
	if len(t.GroupIds) == 0 {
		return t.Text
	}
	parts := []string{}
	for _, id := range t.GroupIds {
		parts = append(parts, strconv.Itoa(id))
	}
	return fmt.Sprintf("%s(%s)", t.Text, strings.Join(parts, ","))
}

type Segment struct {
	// Either Tokens or Text will be set
	Tokens []Token `json:"tokens,omitempty"`
	Text   string  `json:"text,omitempty"`
	Index  int     `json:"index"`
}

func (s Segment) ToStr() string {
	parts := []string{}
	for _, token := range s.Tokens {
		parts = append(parts, token.ToStr())
	}
	return strings.Join(parts, " ")
}

func (s Segment) ToStrNoGroups() string {
	if len(s.Text) > 0 {
		return s.Text
	}

	parts := []string{}
	for _, token := range s.Tokens {
		parts = append(parts, token.Text)
	}
	return strings.Join(parts, " ")
}

type Sentence struct {
	Segments []Segment `json:"segments"`
	Index    int       `json:"index"`
}

func (sen Sentence) ToPlainStr() string {
	parts := []string{}
	for _, seg := range sen.Segments {
		parts = append(parts, seg.ToStrNoGroups())
	}
	return strings.Join(parts, " ")
}

type Scene struct {
	Sentences    []Sentence `json:"sentences"`
	MaybeImageId string     `json:"image_id,omitempty"`
}

type Paragraph struct {
	Scenes       []Scene `json:"scenes"`
	MaybeImageId string  `json:"image_id,omitempty"`
}

type Chapter struct {
	MaybeTitle string      `json:"title,omitempty"`
	Paragraphs []Paragraph `json:"paragraphs"`
}

type Story struct {
	Title        string    `json:"title"`
	Chapters     []Chapter `json:"chapters"`
	MaybeImageId string    `json:"image_id,omitempty"`
}

func (s Story) ToPlainStr() string {
	parts := []string{}
	if s.Title != "" {
		parts = append(parts, s.Title)
	}
	for _, chapter := range s.Chapters {
		if chapter.MaybeTitle != "" {
			parts = append(parts, chapter.MaybeTitle)
		}
		for _, paragraph := range chapter.Paragraphs {
			for _, scene := range paragraph.Scenes {
				for _, sentence := range scene.Sentences {
					parts = append(parts, sentence.ToPlainStr())
				}
			}
			parts = append(parts, "")
		}
	}
	return strings.Join(parts, "\n")
}

type StoryMultilingual struct {
	Id            Id               `json:"id"`
	Level         Level            `json:"level"`
	Localizations map[Locale]Story `json:"localizations"`
}

func hasTokenWithNonZeroGroup(tokens []Token) bool {
	for _, token := range tokens {
		if len(token.GroupIds) > 0 {
			return true
		}
	}
	return false
}

func ParseSentence(line string) Sentence {
	line = strings.ReplaceAll(line, "@", "\n")
	var segments []Segment
	for _, segStr := range strings.Split(line, "||") {
		segStr = strings.TrimRight(segStr, " ")
		var tokens []Token
		for _, tokStr := range strings.Split(segStr, " ") {
			if tokStr == "" {
				continue
			}
			match := TOKEN_GROUPS_REGEX.FindString(tokStr)
			var groupIds []int
			if match != "" {
				groupsRaw := strings.Trim(match, "()")
				for _, g := range strings.Split(groupsRaw, ",") {
					g = strings.TrimSpace(g)
					if g != "" && g != "0" {
						idVal, err := strconv.Atoi(g)
						if err != nil {
							panic(fmt.Sprintf("error parsing group id: %v", err))
						}
						groupIds = append(groupIds, idVal)
					}
				}
			}
			text := TOKEN_GROUPS_REGEX.ReplaceAllString(tokStr, "")
			tokens = append(tokens, Token{
				Text:     text,
				GroupIds: groupIds,
			})
		}
		if hasTokenWithNonZeroGroup(tokens) {
			segments = append(segments, Segment{Tokens: tokens})
		} else {
			segments = append(segments, Segment{Text: segStr})
		}
	}
	return Sentence{Segments: segments}
}

func CalculateSentenceAndSegmentIndices(story *StoryMultilingual) {
	for locale, localization := range story.Localizations {
		sentenceIdx := 0
		segmentIdx := 0
		for chapterIdx := range localization.Chapters {
			chapter := &localization.Chapters[chapterIdx]
			for pIdx := range chapter.Paragraphs {
				paragraph := &chapter.Paragraphs[pIdx]
				for sIdx := range paragraph.Scenes {
					scene := &paragraph.Scenes[sIdx]
					for senIdx := range scene.Sentences {
						sentence := &scene.Sentences[senIdx]
						sentence.Index = sentenceIdx
						sentenceIdx++
						for segIdx := range sentence.Segments {
							sentence.Segments[segIdx].Index = segmentIdx
							segmentIdx++
						}
					}
				}
			}
		}
		story.Localizations[locale] = localization
	}
}

func contains(slice []string, val string) bool {
	for _, v := range slice {
		if v == val {
			return true
		}
	}
	return false
}

func extractStringArray(params map[string]interface{}, key string) ([]string, error) {
	val, exists := params[key]
	if !exists {
		return nil, errors.New("key not found in map")
	}

	// Assert the value as []interface{}
	interfaceArray, ok := val.([]interface{})
	if !ok {
		return nil, errors.New("value is not an array")
	}

	// Convert []interface{} to []string
	stringArray := make([]string, len(interfaceArray))
	for i, v := range interfaceArray {
		str, ok := v.(string)
		if !ok {
			return nil, errors.New("array contains a non-string value")
		}
		stringArray[i] = str
	}

	return stringArray, nil
}

func extractOptionalStringArray(params map[string]interface{}, key string) ([]string, error) {
	_, exists := params[key]
	if !exists {
		return nil, nil
	}
	return extractStringArray(params, key)
}

func extractOptionalString(params map[string]interface{}, key string) (string, error) {
	val, exists := params[key]
	if !exists {
		return "", nil
	}
	str, ok := val.(string)
	if !ok {
		return "", errors.New("value is not a string")
	}
	return str, nil
}

func Parse(lines []string) StoryMultilingual {
	var storyId *Id
	var level *Level
	var locales []Locale
	currentLocaleIdx := 0
	currentLineIdx := 0
	currentLine := ""

	var stories []Story

	nextStoryLocaleIdx := func() int {
		currentLocaleIdx = (currentLocaleIdx + 1) % len(locales)
		return currentLocaleIdx
	}

	panicAtLine := func(msg string) {
		panic(fmt.Sprintf("%s at line %d: %s", msg, currentLineIdx, currentLine))
	}

	startStories := func(titles []string, maybeImageId string) {
		if len(titles) != len(locales) {
			panicAtLine("titles length mismatch with locales")
		}
		for _, title := range titles {
			stories = append(stories, Story{
				Title:        title,
				Chapters:     []Chapter{},
				MaybeImageId: maybeImageId,
			})
		}
	}

	ensureChapters := func() {
		for storyIdx := range stories {
			if len(stories[storyIdx].Chapters) == 0 {
				stories[storyIdx].Chapters = append(stories[storyIdx].Chapters, Chapter{Paragraphs: []Paragraph{}})
			}
		}
	}

	startChapter := func(titles []string) {
		if len(titles) != len(stories) {
			panicAtLine("chapter titles length mismatch")
		}
		for i, title := range titles {
			stories[i].Chapters = append(stories[i].Chapters, Chapter{
				MaybeTitle: title,
				Paragraphs: []Paragraph{},
			})
		}
	}

	ensureParagraphs := func() {
		ensureChapters()
		for storyIdx := range stories {
			chapter := &stories[storyIdx].Chapters[len(stories[storyIdx].Chapters)-1]
			if len(chapter.Paragraphs) == 0 {
				chapter.Paragraphs = append(chapter.Paragraphs, Paragraph{Scenes: []Scene{}})
			}
		}
	}

	startParagraph := func(maybeImageId string) {
		ensureChapters()
		for storyIdx := range stories {
			chapter := &stories[storyIdx].Chapters[len(stories[storyIdx].Chapters)-1]
			chapter.Paragraphs = append(chapter.Paragraphs, Paragraph{Scenes: []Scene{}, MaybeImageId: maybeImageId})
		}
	}

	ensureScenes := func() {
		ensureParagraphs()
		for storyIdx := range stories {
			chapter := &stories[storyIdx].Chapters[len(stories[storyIdx].Chapters)-1]
			p := &chapter.Paragraphs[len(chapter.Paragraphs)-1]
			if len(p.Scenes) == 0 {
				p.Scenes = append(p.Scenes, Scene{Sentences: []Sentence{}})
			}
		}
	}

	startScene := func(maybeImageId string) {
		ensureParagraphs()
		for storyIdx := range stories {
			chapter := &stories[storyIdx].Chapters[len(stories[storyIdx].Chapters)-1]
			p := &chapter.Paragraphs[len(chapter.Paragraphs)-1]
			p.Scenes = append(p.Scenes, Scene{Sentences: []Sentence{}, MaybeImageId: maybeImageId})
		}
	}

	var lastTag string
	for currentLineIdx, currentLine = range lines {
		line := strings.TrimSpace(currentLine)
		if strings.HasPrefix(line, "#") || line == "" {
			continue
		}

		if storyId == nil {
			parts := strings.SplitN(line, " ", 2)
			if len(parts) < 2 {
				panicAtLine("Invalid first line, expected 'STORY_ID LEVEL'")
			}
			sid, lvl := parts[0], parts[1]
			if !contains(LEVELS, lvl) {
				panicAtLine("Invalid language proficiency level")
			}
			storyId = &sid
			level = &lvl
			continue
		}

		if len(locales) == 0 {
			locales = strings.Split(line, ",")
			for i := range locales {
				locales[i] = strings.TrimSpace(locales[i])
			}
			continue
		}

		var tag string
		if strings.HasPrefix(line, "/") {
			tag = strings.Split(line, " ")[0]
		}

		if tag != lastTag && currentLocaleIdx != 0 {
			panicAtLine("locale index not reset when tag changed")
		}
		lastTag = tag

		value := strings.TrimSpace(line[len(tag):])
		params := map[string]interface{}{}
		if tag != "" && value != "" {
			err := json.Unmarshal([]byte(value), &params)
			if err != nil {
				panicAtLine(fmt.Sprintf("error parsing json params: %v", err))
			}
		}

		imageId, err := extractOptionalString(params, "image_id")
		if err != nil {
			panicAtLine(fmt.Sprintf("error parsing 'image_id': %v", err))
		}

		titles, err := extractOptionalStringArray(params, "titles")
		if err != nil {
			panicAtLine(fmt.Sprintf("error parsing 'titles': %v", err))
		}

		switch tag {
		case "/t":
			if titles == nil {
				panicAtLine("Missing titles")
			}
			startStories(titles, imageId)
		case "/s":
			startScene(imageId)
		case "/p":
			startParagraph(imageId)
		case "/c":
			if titles == nil {
				titles = make([]string, len(locales))
			}
			startChapter(titles)
		case "":
			// a sentence line
			ensureScenes()
			sen := ParseSentence(value)
			currentChapter := &stories[currentLocaleIdx].Chapters[len(stories[currentLocaleIdx].Chapters)-1]
			currentParagraph := &currentChapter.Paragraphs[len(currentChapter.Paragraphs)-1]
			currentScene := &currentParagraph.Scenes[len(currentParagraph.Scenes)-1]
			currentScene.Sentences = append(currentScene.Sentences, sen)
			nextStoryLocaleIdx()
		default:
			panicAtLine(fmt.Sprintf("Unknown tag: %s", tag))
		}
	}

	if currentLocaleIdx != 0 {
		panic("final locale index not zero")
	}
	if storyId == nil || level == nil {
		panic("story_id or level not defined")
	}

	locMap := make(map[string]Story)
	for i, loc := range locales {
		locMap[loc] = stories[i]
	}
	story := StoryMultilingual{
		Id:            *storyId,
		Level:         *level,
		Localizations: locMap,
	}
	CalculateSentenceAndSegmentIndices(&story)
	return story
}

func Test() {
	data := `
beneath-peeling-paint-c1 B2-C1
en, de, ru
/t { "titles": ["Beneath", "Unter", "Под"], "image_id": "4" } 
/c { "titles": ["Chapter I", "Kapitel I", "Глава I"] }
/s 
The phone rang(1) || slicing through.
Das schrille Klingeln(1) || durchbrach.
Зазвонил(2) телефон || нарушил.

It was Beatrice.
Es war Beatrice.
Это была Беатрис.

/s
News traveled fast in that small community.
Nachrichten verbreiteten sich schnell in dieser kleinen Gemeinde.
В этой маленькой общине новости распространялись быстро.

/p
/s
The words echoed inside her.
Die Worte hallten in ihr wider.
Слова эхом отозвались в ее душе́.

It wasn't just about the building.
Es ging nicht nur um das Gebäude.
Дело было не просто в здании.

/s
She realized with a pang.
Mit einem Stich im Herzen.
С болью она поняла, что.

/s
Driven by a mix of longing.
Angetrieben von einer Mischung aus Sehnsucht.
Движимая смесью тоски и странного чувства долга.

/c { "titles": ["Chapter II", "Kapitel II", "Глава II"] }
The paint.
Weiße Farbe.
Краска.

Weeds sprouted brazenly.
Unkraut spross dreist.
Сквозь трещины в тротуаре.

Yet, it was the silence that was most striking.
Doch am auffälligsten war die Stille.
Но больше всего поражала тишина.

/p { "image_id": "18" }
Inside, she didn't need to see.
Drinnen musste sie sich nicht auf ihre Augen verlassen.
Внутри ей не нужно было полагаться на зрение.

/s
She closed her eyes, seeing herself.
Sie schloss die Augen und sah sich selbst.
Она закрыла глаза, видя себя.

/p { "image_id": "8" }
Finally, she stood at the threshold.
Schließlich stand sie an der Schwelle.
Наконец, она остановилась у порога.
`
	lines := strings.Split(string(data), "\n")
	parsed := Parse(lines)
	out, err := json.MarshalIndent(parsed, "", "  ")
	if err != nil {
		panic(err)
	}
	fmt.Println(string(out))
}
